### PR TITLE
Allow download of files from within a valid partial form session

### PR DESCRIPTION
**Pull request by scott-nz**

### CODE REVIEW

File: src/controllers/PartialFileDownloadController.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * @package Firesphere\PartialUserforms\Controllers
+ */
+class PartialFileDownloadController extends UserDefinedFormController
+{
+    use PartialSubmissionValidationTrait;
+    /**
+     * @var array
+     */
+    private static $url_handlers = [
+        '$Key/$Token/$ID' => 'partialfiledownload',
```
Comment (scott-nz) [inline src/controllers/PartialFileDownloadController.php]: update $ID

File: src/controllers/PartialUserFormController.php [hunk 2]
```diff
@@ -203,13 +142,23 @@ protected function populateData($form, $partial)
...
+            $request = $this->getRequest();
             $file = $upload->UploadedFile();
             $fileAttributes = ['PartialID' => $partial->ID, 'FileID' => $file->ID];
+
+            // Generate a unique download path that is specific to the current session, partial submission and field
+            $linkSrc = sprintf(
+                'partialfiledownload/%s/%s/%s',
+                $request->param('Key'),
+            $request->param('Token'),
+            $upload->Name
```
Comment (scott-nz) [inline src/controllers/PartialUserFormController.php]: better var name

Comment (scott-nz) [reply src/controllers/PartialUserFormController.php]: fix indenting

File: src/controllers/PartialFileDownloadController.php (hunk 1 above)
Comment (scott-nz) [reply src/controllers/PartialFileDownloadController.php]: fix type hints

File: src/controllers/PartialFileDownloadController.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    public function partialfiledownload(HTTPRequest $request)
+    {
+        /** @var PartialFormSubmission $partial */
+        $partial = $this->validateToken($request);
+        $page = $partial->UserDefinedForm();
+
+        // Check if form is locked
+        if (static::isLockedOut()) {
+            return $this->redirect($page->link('overview'));
+        } else {
```
Comment (scott-nz) [inline src/controllers/PartialFileDownloadController.php]: else is not needed

File: src/controllers/PartialFileDownloadController.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            return $this->httpError(404);
+        }
+
+        $file = $upload->UploadedFile();
+
+        $filesize = $file->getAbsoluteSize();
+        $mime = $file->getMimeType();
+        $stream = $file->getStream();
+
+        if ($stream != null || gettype($stream) == "resource") {
```
Comment (scott-nz) [inline src/controllers/PartialFileDownloadController.php]: strict type this

File: src/controllers/PartialFileDownloadController.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            header('Content-Type: ' . $mime);
+            header('Content-Length: ' . $filesize, false);
+            if (!empty($mime) && $mime != "text/html") {
```
Comment (scott-nz) [inline src/controllers/PartialFileDownloadController.php]: strict type

File: src/controllers/PartialFileDownloadController.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            // Claim the form session
+            PartialSubmissionController::reloadSession($request->getSession(), $partial->ID);
+        }
+
+        $fieldName = $request->param('FieldName');
+
+        // Files are stored in draft state for security reasons.
+        // At this point the partial form submission session has been validated, we can safely retrieve the file now
+        // by temporarily forcing the draft state
+        $upload =  Versioned::withVersionedMode(function () use ($fieldName, $partial) {
```
Comment (scott-nz) [inline src/controllers/PartialFileDownloadController.php]: $uploadField

File: src/controllers/PartialFileDownloadController.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
+<?php
```
Comment (dannidickson) [inline src/controllers/PartialFileDownloadController.php:1]: ⛏️ 👾 there is a cheeky spaces here we can remove.

**Review by dannidickson — APPROVED** [review]
Looks good to me. There is a couple of whitespace issues - I wont block approval though.

**Pull request merged**